File: MacRaptorHAB/RaptorHabGS/GPSManager.swift
//
//  GPSManager.swift
//  RaptorHabGS
//
//  Manages external USB GPS devices for ground station position
//  Parses NMEA 0183 sentences from serial GPS receivers
//

import Foundation
import CoreLocation
import Combine
import Darwin

// MARK: - GPS Data Model

struct GPSPosition: Equatable {
    var latitude: Double
    var longitude: Double
    var altitude: Double  // meters
    var speed: Double     // m/s
    var course: Double    // degrees true
    var satellites: Int
    var fixQuality: GPSFixQuality
    var hdop: Double
    var timestamp: Date
    
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
    
    var isValid: Bool {
        fixQuality != .noFix && latitude != 0 && longitude != 0
    }
    
    static func == (lhs: GPSPosition, rhs: GPSPosition) -> Bool {
        lhs.latitude == rhs.latitude && lhs.longitude == rhs.longitude && lhs.altitude == rhs.altitude
    }
}

enum GPSFixQuality: Int, CustomStringConvertible {
    case noFix = 0
    case gpsFix = 1
    case dgpsFix = 2
    case ppsFix = 3
    case rtkFixed = 4
    case rtkFloat = 5
    case estimated = 6
    case manual = 7
    case simulation = 8
    
    var description: String {
        switch self {
        case .noFix: return "No Fix"
        case .gpsFix: return "GPS"
        case .dgpsFix: return "DGPS"
        case .ppsFix: return "PPS"
        case .rtkFixed: return "RTK Fixed"
        case .rtkFloat: return "RTK Float"
        case .estimated: return "Estimated"
        case .manual: return "Manual"
        case .simulation: return "Simulation"
        }
    }
}

// MARK: - Bearing and Distance

struct BearingDistance {
    var bearing: Double      // degrees true (0-360)
    var distance: Double     // meters
    var elevation: Double    // degrees (positive = look up)
    var altitudeDiff: Double // meters
    
    var bearingCardinal: String {
        let directions = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
                          "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"]
        let index = Int((bearing + 11.25) / 22.5) % 16
        return directions[index]
    }
    
    var distanceFormatted: String {
        if distance < 1000 {
            return String(format: "%.0f m", distance)
        } else {
            return String(format: "%.2f km", distance / 1000)
        }
    }
    
    var distanceMiles: String {
        let miles = distance / 1609.344
        if miles < 1 {
            let feet = distance * 3.28084
            return String(format: "%.0f ft", feet)
        }
        return String(format: "%.2f mi", miles)
    }
}

// MARK: - GPS Manager

class GPSManager: ObservableObject {
    
    static let shared = GPSManager()
    
    // Published properties
    @Published var isConnected = false
    @Published var currentPosition: GPSPosition?
    @Published var bearingToPayload: BearingDistance?
    @Published var availablePorts: [String] = []
    @Published var selectedPort: String = ""
    @Published var statusMessage: String = "Disconnected"
    @Published var baudRate: Int = 9600
    @Published var lastNMEAReceived: Date?  // Track when we last got valid NMEA data
    
    // Check if NMEA data is actively being received (within last 3 seconds)
    var isReceivingData: Bool {
        guard let lastReceived = lastNMEAReceived else { return false }
        return Date().timeIntervalSince(lastReceived) < 3.0
    }
    
    // Serial port
    private var fileDescriptor: Int32 = -1
    private var readThread: Thread?
    private var shouldRead = false
    private var receiveBuffer = ""
    
    // NMEA parsing state
    private var pendingPosition = GPSPosition(
        latitude: 0, longitude: 0, altitude: 0, speed: 0, course: 0,
        satellites: 0, fixQuality: .noFix, hdop: 99.9, timestamp: Date()
    )
    
    private init() {
        refreshPorts()
    }
    
    deinit {
        disconnect()
    }
    
    // MARK: - Port Management
    
    func refreshPorts() {
        var ports: [String] = []
        
        if let devices = try? FileManager.default.contentsOfDirectory(atPath: "/dev") {
            for device in devices {
                // USB serial adapters commonly used for GPS
                if device.hasPrefix("cu.usbserial") ||
                   device.hasPrefix("cu.usbmodem") ||
                   device.hasPrefix("cu.SLAB_USBtoUART") ||  // CP210x
                   device.hasPrefix("cu.wchusbserial") ||    // CH340
                   device.hasPrefix("cu.PL2303") ||          // Prolific
                   device.contains("GPS") ||
                   device.contains("gps") {
                    ports.append("/dev/" + device)
                }
            }
        }
        
        availablePorts = ports.sorted()
        
        // Auto-select if one port found
        if selectedPort.isEmpty && ports.count == 1 {
            selectedPort = ports[0]
        }
    }
    
    // MARK: - Connection
    
    func connect() {
        guard !selectedPort.isEmpty else {
            statusMessage = "No port selected"
            return
        }
        
        disconnect()
        
        // Open serial port with nonblocking to prevent hanging on open
        fileDescriptor = open(selectedPort, O_RDWR | O_NOCTTY | O_NONBLOCK)
        
        guard fileDescriptor != -1 else {
            statusMessage = "Failed to open port"
            return
        }
        
        // Configure port - use cfmakeraw for simplest raw mode
        var options = termios()
        tcgetattr(fileDescriptor, &options)
        
        // Set to raw mode (minimal processing)
        cfmakeraw(&options)
        
        // Set baud rate (may be ignored by USB CDC devices)
        let speed: speed_t
        switch baudRate {
        case 4800: speed = speed_t(B4800)
        case 9600: speed = speed_t(B9600)
        case 19200: speed = speed_t(B19200)
        case 38400: speed = speed_t(B38400)
        case 57600: speed = speed_t(B57600)
        case 115200: speed = speed_t(B115200)
        default: speed = speed_t(B9600)
        }
        cfsetspeed(&options, speed)
        
        // Enable receiver and set local mode
        options.c_cflag |= UInt(CLOCAL | CREAD)
        
        // Apply settings
        tcsetattr(fileDescriptor, TCSANOW, &options)
        
        // Flush any pending data
        tcflush(fileDescriptor, TCIOFLUSH)
        
        isConnected = true
        let portName = selectedPort.components(separatedBy: "/").last ?? selectedPort
        statusMessage = "Connected: \(portName)"
        
        // Start reading thread
        shouldRead = true
        readThread = Thread { [weak self] in
            self?.readLoop()
        }
        readThread?.qualityOfService = .userInitiated
        readThread?.start()
    }
    
    func disconnect() {
        shouldRead = false
        readThread?.cancel()
        readThread = nil
        
        if fileDescriptor != -1 {
            close(fileDescriptor)
            fileDescriptor = -1
        }
        
        isConnected = false
        statusMessage = "Disconnected"
        currentPosition = nil
        bearingToPayload = nil
        lastNMEAReceived = nil
    }
    
    // MARK: - Serial Reading
    
    private func readLoop() {
        var buffer = [UInt8](repeating: 0, count: 512)
        
        while shouldRead && fileDescriptor != -1 {
            let bytesRead = read(fileDescriptor, &buffer, buffer.count)
            
            if bytesRead > 0 {
                // Process as raw bytes directly
                let data = Data(buffer[0..<bytesRead])
                processBytes(data)
                
            } else if bytesRead == 0 {
                // Timeout or EOF - normal for VTIME (50ms = 20 reads/sec max)
                Thread.sleep(forTimeInterval: 0.05)
            } else {
                // Error
                let err = errno
                if err == EAGAIN || err == EWOULDBLOCK {
                    // No data available in non-blocking mode (50ms = 20 reads/sec max)
                    Thread.sleep(forTimeInterval: 0.05)
                } else if err != EINTR {
                    DispatchQueue.main.async { [weak self] in
                        self?.disconnect()
                        self?.statusMessage = "Read error: \(err)"
                    }
                    break
                }
            }
        }
    }
    
    private var byteBuffer = Data()
    
    private func processData(_ data: String) {
        // Convert string back to data and process as bytes
        if let newData = data.data(using: .ascii) {
            processBytes(newData)
        }
    }
    
    private func processBytes(_ data: Data) {
        byteBuffer.append(data)
        
        // Look for complete NMEA sentences (end with CR LF or just LF)
        while true {
            // Find $ which starts an NMEA sentence
            guard let dollarIndex = byteBuffer.firstIndex(of: 0x24) else { // 0x24 = '$'
                byteBuffer.removeAll()
                break
            }
            
            // Remove any garbage before the $
            if dollarIndex > byteBuffer.startIndex {
                byteBuffer.removeSubrange(byteBuffer.startIndex..<dollarIndex)
            }
            
            // Look for LF (0x0A) which ends the sentence
            guard let lfIndex = byteBuffer.firstIndex(of: 0x0A) else { // 0x0A = '\n'
                break
            }
            
            // Extract the sentence (without CR LF)
            var endIndex = lfIndex
            if endIndex > byteBuffer.startIndex {
                let prevIndex = byteBuffer.index(before: endIndex)
                if byteBuffer[prevIndex] == 0x0D { // 0x0D = '\r'
                    endIndex = prevIndex
                }
            }
            
            let sentenceData = byteBuffer[byteBuffer.startIndex..<endIndex]
            if let sentence = String(data: sentenceData, encoding: .ascii) {
                parseNMEASentence(sentence)
            }
            
            // Remove processed sentence from buffer (including CR LF)
            let removeEnd = byteBuffer.index(after: lfIndex)
            byteBuffer.removeSubrange(byteBuffer.startIndex..<removeEnd)
        }
        
        // Prevent buffer overflow
        if byteBuffer.count > 4096 {
            byteBuffer.removeAll()
        }
    }
    
    // MARK: - NMEA Parsing
    
    private func parseNMEASentence(_ sentence: String) {
        // Validate checksum first
        guard validateChecksum(sentence) else { return }
        
        // Update activity timestamp on valid NMEA sentence
        DispatchQueue.main.async { [weak self] in
            self?.lastNMEAReceived = Date()
        }
        
        let parts = sentence.components(separatedBy: "*")[0]
        let fields = parts.components(separatedBy: ",")
        guard fields.count > 0 else { return }
        
        let type = fields[0]
        
        if type.hasSuffix("GGA") {
            parseGGA(fields)
        } else if type.hasSuffix("RMC") {
            parseRMC(fields)
        } else if type.hasSuffix("VTG") {
            parseVTG(fields)
        }
    }
    
    private func validateChecksum(_ sentence: String) -> Bool {
        guard sentence.hasPrefix("$"),
              let asteriskIndex = sentence.firstIndex(of: "*") else { return false }
        
        let data = sentence[sentence.index(after: sentence.startIndex)..<asteriskIndex]
        let checksumStr = String(sentence[sentence.index(after: asteriskIndex)...]).prefix(2)
        
        var checksum: UInt8 = 0
        for char in data.utf8 {
            checksum ^= char
        }
        
        return String(format: "%02X", checksum) == checksumStr.uppercased()
    }
    
    // GGA - GPS Fix Data
    private func parseGGA(_ fields: [String]) {
        guard fields.count >= 15 else { return }
        
        let lat = parseLatitude(fields[2], direction: fields[3])
        let lon = parseLongitude(fields[4], direction: fields[5])
        let quality = Int(fields[6]) ?? 0
        let satellites = Int(fields[7]) ?? 0
        let hdop = Double(fields[8]) ?? 99.9
        let altitude = Double(fields[9]) ?? 0
        
        pendingPosition.latitude = lat
        pendingPosition.longitude = lon
        pendingPosition.altitude = altitude
        pendingPosition.satellites = satellites
        pendingPosition.fixQuality = GPSFixQuality(rawValue: quality) ?? .noFix
        pendingPosition.hdop = hdop
        pendingPosition.timestamp = Date()
        
        if pendingPosition.isValid {
            publishPosition()
        }
    }
    
    // RMC - Recommended Minimum
    private func parseRMC(_ fields: [String]) {
        guard fields.count >= 12, fields[2] == "A" else { return }
        
        let lat = parseLatitude(fields[3], direction: fields[4])
        let lon = parseLongitude(fields[5], direction: fields[6])
        let speedKnots = Double(fields[7]) ?? 0
        let course = Double(fields[8]) ?? 0
        
        pendingPosition.latitude = lat
        pendingPosition.longitude = lon
        pendingPosition.speed = speedKnots * 0.514444
        pendingPosition.course = course
        pendingPosition.timestamp = Date()
        
        if pendingPosition.fixQuality == .noFix {
            pendingPosition.fixQuality = .gpsFix
        }
        
        if pendingPosition.isValid {
            publishPosition()
        }
    }
    
    // VTG - Course Over Ground
    private func parseVTG(_ fields: [String]) {
        guard fields.count >= 9 else { return }
        
        if let course = Double(fields[1]), course > 0 {
            pendingPosition.course = course
        }
        if let speedKmh = Double(fields[7]) {
            pendingPosition.speed = speedKmh / 3.6
        }
    }
    
    private func parseLatitude(_ value: String, direction: String) -> Double {
        guard value.count >= 4 else { return 0 }
        let degrees = Double(value.prefix(2)) ?? 0
        let minutes = Double(value.dropFirst(2)) ?? 0
        var result = degrees + minutes / 60.0
        if direction == "S" { result = -result }
        return result
    }
    
    private func parseLongitude(_ value: String, direction: String) -> Double {
        guard value.count >= 5 else { return 0 }
        let degrees = Double(value.prefix(3)) ?? 0
        let minutes = Double(value.dropFirst(3)) ?? 0
        var result = degrees + minutes / 60.0
        if direction == "W" { result = -result }
        return result
    }
    
    private func publishPosition() {
        let pos = pendingPosition
        DispatchQueue.main.async { [weak self] in
            self?.currentPosition = pos
        }
    }
    
    // MARK: - Bearing/Distance Calculation
    
    func updateBearing(toLatitude lat: Double, toLongitude lon: Double, toAltitude alt: Double) {
        guard let pos = currentPosition, pos.isValid else {
            DispatchQueue.main.async { [weak self] in
                self?.bearingToPayload = nil
            }
            return
        }
        
        let bearing = calculateBearing(from: pos.coordinate,
                                        to: CLLocationCoordinate2D(latitude: lat, longitude: lon))
        let distance = calculateDistance(from: pos.coordinate,
                                          to: CLLocationCoordinate2D(latitude: lat, longitude: lon))
        let altDiff = alt - pos.altitude
        let elevation = atan2(altDiff, distance) * 180 / .pi
        
        DispatchQueue.main.async { [weak self] in
            self?.bearingToPayload = BearingDistance(
                bearing: bearing,
                distance: distance,
                elevation: elevation,
                altitudeDiff: altDiff
            )
        }
    }
    
    private func calculateBearing(from: CLLocationCoordinate2D, to: CLLocationCoordinate2D) -> Double {
        let lat1 = from.latitude * .pi / 180
        let lat2 = to.latitude * .pi / 180
        let dLon = (to.longitude - from.longitude) * .pi / 180
        
        let y = sin(dLon) * cos(lat2)
        let x = cos(lat1) * sin(lat2) - sin(lat1) * cos(lat2) * cos(dLon)
        
        let bearing = atan2(y, x) * 180 / .pi
        return (bearing + 360).truncatingRemainder(dividingBy: 360)
    }
    
    private func calculateDistance(from: CLLocationCoordinate2D, to: CLLocationCoordinate2D) -> Double {
        let R = 6371000.0  // Earth radius in meters
        
        let lat1 = from.latitude * .pi / 180
        let lat2 = to.latitude * .pi / 180
        let dLat = (to.latitude - from.latitude) * .pi / 180
        let dLon = (to.longitude - from.longitude) * .pi / 180
        
        let a = sin(dLat/2) * sin(dLat/2) +
                cos(lat1) * cos(lat2) * sin(dLon/2) * sin(dLon/2)
        let c = 2 * atan2(sqrt(a), sqrt(1-a))
        
        return R * c
    }
}

File: groundstation/python/raptorhabgs/core/gps_manager.py
"""
GPS Manager for ground station position.
Parses NMEA data from a serial GPS receiver.
"""

import serial
import serial.tools.list_ports
import math
from threading import Thread, Event
from typing import Optional, List
from datetime import datetime

from PyQt6.QtCore import QObject, pyqtSignal

from .telemetry import GPSPosition, BearingDistance


class GPSManager(QObject):
    """
    Manages GPS serial connection and NMEA parsing.
    
    Signals:
        position_updated: Emitted when GPS position is updated
        bearing_updated: Emitted when bearing to target is calculated
        connected: Emitted on successful connection
        disconnected: Emitted on disconnection
        error: Emitted on error
    """
    
    position_updated = pyqtSignal(object)  # GPSPosition
    bearing_updated = pyqtSignal(object)   # BearingDistance
    connected = pyqtSignal()
    disconnected = pyqtSignal()
    error = pyqtSignal(str)
    
    def __init__(self, parent=None):
        super().__init__(parent)
        
        self.serial: Optional[serial.Serial] = None
        self.is_connected = False
        
        self._read_thread: Optional[Thread] = None
        self._stop_event = Event()
        
        self._byte_buffer = bytearray()
        
        self.current_position: Optional[GPSPosition] = None
        self.current_bearing: Optional[BearingDistance] = None
        self._pending_position = GPSPosition()
        
        self.status_message = "Disconnected"
    
    @staticmethod
    def list_ports() -> List[str]:
        """List available serial ports."""
        ports = serial.tools.list_ports.comports()
        return [port.device for port in ports]
    
    def connect(self, port: str, baud_rate: int = 9600) -> bool:
        """Connect to GPS serial port."""
        if self.is_connected:
            self.disconnect()
        
        try:
            self.serial = serial.Serial(
                port=port,
                baudrate=baud_rate,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE,
                timeout=0.1,
            )
            
            self.serial.reset_input_buffer()
            self.is_connected = True
            self._byte_buffer.clear()
            self.status_message = "Connected"
            
            # Start read thread
            self._stop_event.clear()
            self._read_thread = Thread(target=self._read_loop, daemon=True)
            self._read_thread.start()
            
            self.connected.emit()
            return True
            
        except Exception as e:
            self.error.emit(f"GPS connection failed: {e}")
            self.status_message = f"Error: {e}"
            return False
    
    def disconnect(self):
        """Disconnect from GPS."""
        if not self.is_connected:
            return
        
        self._stop_event.set()
        if self._read_thread and self._read_thread.is_alive():
            self._read_thread.join(timeout=1.0)
        self._read_thread = None
        
        if self.serial:
            try:
                self.serial.close()
            except:
                pass
            self.serial = None
        
        self.is_connected = False
        self.status_message = "Disconnected"
        self.current_position = None
        
        self.disconnected.emit()
    
    def update_bearing(self, target_lat: float, target_lon: float, target_alt: float = 0):
        """Calculate bearing and distance to target."""
        if not self.current_position or not self.current_position.is_valid:
            return
        
        pos = self.current_position
        bearing = self._calculate_bearing(
            pos.latitude, pos.longitude,
            target_lat, target_lon
        )
        
        distance = self._haversine_distance(
            pos.latitude, pos.longitude,
            target_lat, target_lon
        )
        
        # Calculate elevation angle
        if distance > 0:
            alt_diff = target_alt - pos.altitude
            elevation = math.degrees(math.atan2(alt_diff, distance))
        else:
            elevation = 90.0 if target_alt > pos.altitude else 0.0
        
        self.current_bearing = BearingDistance(
            bearing=bearing,
            distance=distance,
            elevation=elevation
        )
        
        self.bearing_updated.emit(self.current_bearing)
    
    def _read_loop(self):
        """Background thread for reading GPS data."""
        while not self._stop_event.is_set() and self.serial:
            try:
                if self.serial.in_waiting > 0:
                    data = self.serial.read(self.serial.in_waiting)
                    if data:
                        self._process_bytes(data)
                else:
                    self._stop_event.wait(0.05)
                    
            except serial.SerialException as e:
                if not self._stop_event.is_set():
                    self.error.emit(f"GPS read error: {e}")
                break
            except Exception:
                pass
    
    def _process_bytes(self, data: bytes):
        """Process received bytes and extract NMEA sentences."""
        self._byte_buffer.extend(data)
        
        while True:
            # Find $ which starts NMEA sentence
            try:
                dollar_idx = self._byte_buffer.index(ord('$'))
            except ValueError:
                self._byte_buffer.clear()
                break
            
            # Remove garbage before $
            if dollar_idx > 0:
                del self._byte_buffer[:dollar_idx]
            
            # Find newline
            try:
                newline_idx = self._byte_buffer.index(ord('\n'))
            except ValueError:
                break
            
            # Extract sentence
            sentence_bytes = bytes(self._byte_buffer[:newline_idx])
            del self._byte_buffer[:newline_idx + 1]
            
            # Clean up and parse
            try:
                sentence = sentence_bytes.decode('ascii').strip()
                if sentence.startswith('$') and self._verify_checksum(sentence):
                    self._parse_nmea(sentence)
            except:
                pass
    
    def _verify_checksum(self, sentence: str) -> bool:
        """Verify NMEA sentence checksum."""
        if '*' not in sentence:
            return False
        
        try:
            data, checksum_str = sentence[1:].split('*')
            expected = int(checksum_str, 16)
            
            calculated = 0
            for char in data:
                calculated ^= ord(char)
            
            return calculated == expected
        except:
            return False
    
    def _parse_nmea(self, sentence: str):
        """Parse NMEA sentence."""
        parts = sentence.split(',')
        if len(parts) < 2:
            return
        
        msg_type = parts[0]
        
        if msg_type in ('$GPGGA', '$GNGGA'):
            self._parse_gga(parts)
        elif msg_type in ('$GPRMC', '$GNRMC'):
            self._parse_rmc(parts)
        elif msg_type in ('$GPGSA', '$GNGSA'):
            self._parse_gsa(parts)
    
    def _parse_gga(self, fields: List[str]):
        """Parse GGA sentence (fix data)."""
        if len(fields) < 15:
            return
        
        try:
            # Latitude
            if fields[2] and fields[3]:
                lat = self._parse_coordinate(fields[2], fields[3])
                self._pending_position.latitude = lat
            
            # Longitude
            if fields[4] and fields[5]:
                lon = self._parse_coordinate(fields[4], fields[5])
                self._pending_position.longitude = lon
            
            # Fix quality
            if fields[6]:
                self._pending_position.fix_quality = int(fields[6])
            
            # Satellites
            if fields[7]:
                self._pending_position.satellites = int(fields[7])
            
            # HDOP
            if fields[8]:
                self._pending_position.hdop = float(fields[8])
            
            # Altitude
            if fields[9]:
                self._pending_position.altitude = float(fields[9])
            
            self._pending_position.timestamp = datetime.now()
            
            # Update position if valid
            if self._pending_position.is_valid:
                self.current_position = GPSPosition(
                    latitude=self._pending_position.latitude,
                    longitude=self._pending_position.longitude,
                    altitude=self._pending_position.altitude,
                    satellites=self._pending_position.satellites,
                    fix_quality=self._pending_position.fix_quality,
                    hdop=self._pending_position.hdop,
                    timestamp=self._pending_position.timestamp
                )
                self.status_message = f"{self.current_position.satellites} satellites"
                self.position_updated.emit(self.current_position)
                
        except Exception as e:
            pass
    
    def _parse_rmc(self, fields: List[str]):
        """Parse RMC sentence (recommended minimum)."""
        if len(fields) < 12:
            return
        
        try:
            # Check validity
            if fields[2] != 'A':
                return
            
            # Latitude
            if fields[3] and fields[4]:
                lat = self._parse_coordinate(fields[3], fields[4])
                self._pending_position.latitude = lat
            
            # Longitude
            if fields[5] and fields[6]:
                lon = self._parse_coordinate(fields[5], fields[6])
                self._pending_position.longitude = lon
                
        except:
            pass
    
    def _parse_gsa(self, fields: List[str]):
        """Parse GSA sentence (DOP and active satellites)."""
        if len(fields) < 18:
            return
        
        try:
            # Fix type: 1=no fix, 2=2D, 3=3D
            if fields[2]:
                fix_type = int(fields[2])
                if fix_type >= 2:
                    self._pending_position.fix_quality = fix_type - 1
            
            # HDOP
            if fields[16]:
                self._pending_position.hdop = float(fields[16])
                
        except:
            pass
    
    def _parse_coordinate(self, value: str, direction: str) -> float:
        """Parse NMEA coordinate (DDMM.MMMM or DDDMM.MMMM)."""
        if not value:
            return 0.0
        
        # Find decimal point
        dot_idx = value.index('.')
        
        # Degrees are before the last 2 digits before decimal
        degrees = int(value[:dot_idx - 2])
        minutes = float(value[dot_idx - 2:])
        
        coord = degrees + minutes / 60.0
        
        if direction in ('S', 'W'):
            coord = -coord
        
        return coord
    
    @staticmethod
    def _calculate_bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """Calculate bearing from point 1 to point 2."""
        lat1_rad = math.radians(lat1)
        lat2_rad = math.radians(lat2)
        delta_lon = math.radians(lon2 - lon1)
        
        x = math.sin(delta_lon) * math.cos(lat2_rad)
        y = (math.cos(lat1_rad) * math.sin(lat2_rad) -
             math.sin(lat1_rad) * math.cos(lat2_rad) * math.cos(delta_lon))
        
        bearing = math.degrees(math.atan2(x, y))
        return (bearing + 360) % 360
    
    @staticmethod
    def _haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """Calculate distance between two points using Haversine formula."""
        R = 6371000  # Earth radius in meters
        
        phi1 = math.radians(lat1)
        phi2 = math.radians(lat2)
        delta_phi = math.radians(lat2 - lat1)
        delta_lambda = math.radians(lon2 - lon1)
        
        a = (math.sin(delta_phi / 2) ** 2 +
             math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2) ** 2)
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        
        return R * c
